File: EnglishSentence/EnglishSentence/SentenceModels.swift
import Foundation

// MARK: - CourseUnit
struct CourseUnit: Codable {
    let unitName: String
    let description: String
    let sentences: [Sentence]

    enum CodingKeys: String, CodingKey {
        case unitName = "unit_name"
        case description
        case sentences
    }
}

// MARK: - Sentence
struct Sentence: Codable {
    let id: String
    let sentenceInfo: SentenceInfo
    let analysis: [WordAnalysis]

    enum CodingKeys: String, CodingKey {
        case id
        case sentenceInfo = "sentence_info"
        case analysis
    }
}

// MARK: - SentenceInfo
struct SentenceInfo: Codable {
    let original: String
    let translation: String
    let sentencePattern: String
    let patternCode: String
    let difficultyLevel: Int
    let tags: [String]

    enum CodingKeys: String, CodingKey {
        case original
        case translation
        case sentencePattern = "sentence_pattern"
        case patternCode = "pattern_code"
        case difficultyLevel = "difficulty_level"
        case tags
    }
}

// MARK: - WordAnalysis
struct WordAnalysis: Codable {
    let word: String
    let baseForm: String
    let range: [Int]
    let tag: String
    let type: String
    let chineseDefinition: String
    let explanation: String
    let phonetics: Phonetics

    enum CodingKeys: String, CodingKey {
        case word
        case baseForm = "base_form"
        case range
        case tag
        case type
        case chineseDefinition = "chinese_definition"
        case explanation
        case phonetics
    }
}

// MARK: - Phonetics
struct Phonetics: Codable {
    let uk: String
    let us: String
    let audioUk: String
    let audioUs: String

    enum CodingKeys: String, CodingKey {
        case uk
        case us
        case audioUk = "audio_uk"
        case audioUs = "audio_us"
    }
}

File: EnglishSentence/AndroidEnglishSentence/app/src/main/java/com/example/englishsentence/data/model/SentenceModels.kt
package com.example.englishsentence.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class CourseUnit(
    @SerialName("unit_name") val unitName: String,
    val description: String,
    val category: String? = null,
    val code: String? = null,
    val sentences: List<Sentence>,
) {
    val id: String get() = code ?: unitName
}

@Serializable
data class Sentence(
    val id: String,
    @SerialName("sentence_info") val sentenceInfo: SentenceInfo,
    val analysis: List<WordAnalysis>,
)

@Serializable
data class SentenceInfo(
    val original: String,
    val translation: String,
    @SerialName("sentence_pattern") val sentencePattern: String,
    @SerialName("pattern_code") val patternCode: String,
    @SerialName("difficulty_level") val difficultyLevel: Int,
    val tags: List<String> = emptyList(),
)

@Serializable
data class WordAnalysis(
    val word: String,
    @SerialName("base_form") val baseForm: String,
    val range: List<Int> = emptyList(),
    val tag: String,
    val type: String,
    @SerialName("chinese_definition") val chineseDefinition: String,
    val explanation: String,
    val phonetics: Phonetics = Phonetics(),
)

@Serializable
data class Phonetics(
    val uk: String? = null,
    val us: String? = null,
    @SerialName("audio_uk") val audioUk: String? = null,
    @SerialName("audio_us") val audioUs: String? = null,
)

@Serializable
data class CourseManifest(
    val sections: List<ManifestSection>,
)

@Serializable
data class ManifestSection(
    val id: String,
    val title: String,
    val directory: String,
    val files: List<String>,
)

data class CourseSection(
    val categoryName: String,
    val courses: List<CourseUnit>,
)
